source("function.r")

#need to take in parameter
## passed in from the command prompt.
parseCommandArgs()
## Will disply the default values on the first run,
print(seed)
print(n_rep)
print(r)
print(sigma2)
print(n_family) #number of family
n_pair=n_family #number of pair for case-control
print(p_dis) #prevelance
print(null)
print(family_strct)
print(method)

## ... your simualtion code
##the founder's haplotypes
#import the haplotypes generated by cosi
haplotype <- read.table("../out_100k_10k_1kb.hap-1", header=F)
colnames(haplotype) <- c("HAP", "CHROM", paste("SNP", 1:(ncol(haplotype)-2), sep=""))
snp <-read.table("../out_100k_10k_1kb.pos-1", header=T)
#make allele 1 is the minor allele, 2 is the common allele
temp.idx <- snp$FREQ1 > snp$FREQ2
temp.freq <- snp$FREQ2
snp$FREQ2[temp.idx] <- snp$FREQ1[temp.idx]
snp$FREQ1[temp.idx] <- temp.freq[temp.idx]
#also change the genotype file
haplotype[,which(temp.idx==T)+2] <- 3 - haplotype[,which(temp.idx==T)+2]

#allele frequency
nrow(snp) #total number of snp
sum(snp$FREQ1 < 0.05) # number of snp with f < 0.05
sum(snp$FREQ1 < 0.01) # number of snp with f < 0.01
sum(snp$FREQ1 == 0.0001) # number of singletons

##assign risk variants and the corresponding effect size (proportional to allele frequency)
# null <- FALSE
n_haplo <- 10000
n_snp <- ncol(haplotype)-2
prevalence <- p_dis
b0_sqrt <- sqrt(prevalence)   #baseline


##gene drop simulation for three generations
family_strct.2g2c <- data.frame(family=c(1,1,1,1), person=c(1,2,3,4), father=c(0,0,1,1), 
                                mother=c(0,0,2,2), sex=c(1,2,1,1), affect=c(1,2,2,2)) #1=male, 2=female, 1=unaffected, 2=affected
family_strct.3g <- data.frame(family=c(1,1,1,1,1,1), person=c(1,2,3,4,5,6), father=c(0,0,1,1,0,4), 
                              mother=c(0,0,2,2,0,5), sex=c(1,2,1,1,2,1), affect=c(1,2,2,2,1,2)) #1=male, 2=female, 1=unaffected, 2=affected
family_strct.2g3c <- data.frame(family=c(1,1,1,1,1), person=c(1,2,3,4,5), father=c(0,0,1,1,1), 
                                mother=c(0,0,2,2,2), sex=c(1,2,1,1,1), affect=c(1,2,2,2,2)) #1=male, 2=female, 1=unaffected, 2=affected

null==FALSE

rep.idx <<- 1
sim_result <- replicate(n_rep, {
  ##generate risk haplotypes
  #   risk.variant.id <- sort(c(sample(which((snp$FREQ1 < 0.05)==T & snp$FREQ1 != 1/n_haplo ), 5), sample(which(snp$FREQ1 == 1/n_haplo), 5))) #randomly pick 5 variants with f < 0.05 and 5 singletons as casual 
  if(sigma2 != 0) {
    #     set.seed(72446) #347281
    #     risk.variant.id <- sort(c(sample(which((snp$FREQ1 < 0.01)==T & snp$FREQ1 != 1/n_haplo ), 5), sample(which(snp$FREQ1 == 1/n_haplo), 5))) #randomly pick 5 variants with f < 0.05 and 5 singletons as casual 
    #     print(risk.variant.id)
    risk.variant.id <- c(3, 5, 15, 21, 24, 26, 28, 32, 43, 50)
    risk.haplo.id <- which(apply(2-haplotype[, risk.variant.id+2], 1, sum)>0)
    (rish.haplo.f <- mean(apply(2-haplotype[, risk.variant.id+2], 1, sum)>0)) #carrier haplotype frequency
    # f= 0.0261
    b <- rep(1, length=(ncol(haplotype)-2))
    if(null==FALSE) b[risk.variant.id] <- abs(log10(snp$FREQ1[risk.variant.id])) #effect size is log10 of its allele frequency
    #calculate the haplotype variants p(A|h)
    haplotype.risk <<- apply(2-haplotype[, -c(1:2)], 1, function(x) prod(b^x)*b0_sqrt)
    #calculate the probability of drawing given the affected status of founder p(h|A)
    # haplotype.draw_prob <- (haplotype.risk %o% haplotype.risk)/n_haplo^2/prevalence
    r <<- mean(haplotype.risk[risk.haplo.id]/b0_sqrt) #mean relative risk
    sigma2 <<- var(haplotype.risk[risk.haplo.id]/b0_sqrt) #mean relative risk
    
  } else{ #var==0 fixed relative risk
    risk.variant.id <- c(3, 5, 15, 21, 24, 26, 28, 32, 43, 50)
    risk.haplo.id <- which(apply(2-haplotype[, risk.variant.id+2], 1, sum)>0)
    (rish.haplo.f <- mean(apply(2-haplotype[, risk.variant.id+2], 1, sum)>0)) #carrier haplotype frequency
    
    haplotype.risk <- rep(1, length=nrow(haplotype))
    #assign mean relative risk calculate the haplotype variants p(A|h)
    if(null==FALSE) haplotype.risk[risk.haplo.id] <- r
    mean(haplotype.risk[risk.haplo.id]) #mean relative risk
    haplotype.risk <<- haplotype.risk*b0_sqrt
  }
  
  print(rep.idx)
  rep.idx <<- rep.idx + 1
  ##if case-control test
  if(family_strct=="cc") {
    case_control_generated <- gene_case_control(n=n_pair)
    case_control.test(data=case_control_generated, f=risk.variant.id)  
  } else{
    ##if family test
    family_generated <- gene_family(family_strct=eval(parse(text=family_strct)), n=n_family) 
    if(family_strct=="family_strct.3g") { #delete the 6th person
      temp.idx <- which(family_generated$data_family$person==5) #take out 5th person
      family_generated$data_family <- family_generated$data_family[-temp.idx, ]
      family_generated$tran_vec <- family_generated$tran_vec[-temp.idx, ]
    }
    family.test(data=family_generated, f=risk.variant.id, method=method)
  }
})

## Write out your results to a csv file
##if case-control test
if(family_strct=="cc") {
  write.csv(data.frame(r=r, sigma2=sigma2, prop.1=sim_result[1,], prop.2=sim_result[2,], p.value=sim_result[3,]),
            paste("res_",r,"_",sigma2,"_",n_pair,"_",seed,"_case_control",".csv",sep=""), row.names=FALSE)
  
} else{
  ##if family test
  write.csv(data.frame(seed=seed, r=r, sigma2=sigma2, method=method, finail.test.stat=sim_result[1,], sum_e=sim_result[2,], se=sim_result[3,], 
                       mean_observed=sim_result[4,], mean_mean=sim_result[5,], mean_var=sim_result[6,],
                       p.value=sim_result[7,], n_info_family=sim_result[8,]),
            paste("res_",r,"_",sigma2,"_",n_family,"_",seed,"_",method,".csv",sep=""), row.names=FALSE)
}


#plot for the results under the null
haplotype <- read.table("out_100k_10k_1kb.hap-1", header=F)
colnames(haplotype) <- c("HAP", "CHROM", paste("SNP", 1:(ncol(haplotype)-2), sep=""))
snp <-read.table("out_100k_10k_1kb.pos-1", header=T)
#make allele 1 is the minor allele, 2 is the common allele
temp.idx <- snp$FREQ1 > snp$FREQ2
temp.freq <- snp$FREQ2
snp$FREQ2[temp.idx] <- snp$FREQ1[temp.idx]
snp$FREQ1[temp.idx] <- temp.freq[temp.idx]
#also change the genotype file
haplotype[,which(temp.idx==T)+2] <- 3 - haplotype[,which(temp.idx==T)+2]
n_haplo <- 10000
null=F
b0_sqrt=sqrt(0.05)

#under the null
alpha=0.05
result.cc <- read.csv("2014-11-25 compare method 1(Seb) and 2(G)_cc.csv", stringsAsFactors=FALSE)
result.cc.null <- read.csv("2014-11-25 compare method 1(Seb) and 2(G)_cc_null.csv", stringsAsFactors=FALSE)
result.2g2c <- read.csv("2014-11-25 compare method 1(Seb) and 2(G)_2g2c.csv", stringsAsFactors=FALSE)
result.2g2c.null <- read.csv("2014-11-25 compare method 1(Seb) and 2(G)_2g2c_null.csv", stringsAsFactors=FALSE)

dev.new(width=6, height=5)
##without allelic heterogeneity
result.cc$r
(freq <- c(mean((result.cc[idx_0, ]$p.value < alpha)), mean((result.2g2c[idx_0, ]$p.value < alpha)), mean((result.2g3c[idx_0, ]$p.value < alpha)), mean((result.3g_5ppl[idx_0, ]$p.value < alpha))))
xx <- barplot(freq, names=c("Case-control", "TRAP 2g2o", "TRAP 2g3o", "TRAP 3g3o"), ylim=c(0,1), ylab="power")
text(x = xx, y = freq, label=freq, pos = 3, cex = 1, col = "red")

dev.new(width=13, height=5)
cex=1.3
par(mfrow=c(1,2), mar=c(3.1, 0.1, 2, 0.1), oma=c(0,3,0,1), lwd=2)
#without allelic heterogeneity
result.cc$r
(freq <- c(mean((result.cc[idx_0, ]$p.value < alpha)), mean((result.2g2c[idx_0, ]$p.value < alpha)), mean((result.2g3c[idx_0, ]$p.value < alpha)), mean((result.3g_5ppl[idx_0, ]$p.value < alpha))))
xx <- barplot(freq, names=c("Case-control", "TRAP 2g2o", "TRAP 2g3o", "TRAP 3g3o"), ylim=c(0,1), ylab="power", cex.axis=cex, cex.lab=cex, cex=cex)
text(x = xx, y = freq, label=round(freq,2), pos = 3, cex = cex, col = "red")
mtext("a. without allelic heterogeneity", 3, 1, adj=0, cex=cex)

#with allelic heterogeneity
result.cc$r
(freq <- c(mean((result.cc[-idx_0, ]$p.value < alpha)), mean((result.2g2c[-idx_0, ]$p.value < alpha)), mean((result.2g3c[-idx_0, ]$p.value < alpha)), mean((result.3g_5ppl[-idx_0, ]$p.value < alpha))))
par(yaxt="n")
xx <- barplot(freq, names=c("Case-control", "TRAP 2g2o", "TRAP 2g3o", "TRAP 3g3o"), ylim=c(0,1), ylab="power", cex.axis=cex, cex.lab=cex, cex=cex)
text(x = xx, y = freq, label=round(freq,2), pos = 3, cex = cex, col = "red")
mtext("b. with allelic heterogeneity", 3, 1, adj=0, cex=cex)

