## ... your simualtion codek
##the founder's haplotypes
#import the haplotypes generated by cosi
haplotype <- read.table("out_100k_10k_1kb.hap-1", header=F)
colnames(haplotype) <- c("HAP", "CHROM", paste("SNP", 1:(ncol(haplotype)-2), sep=""))
snp <-read.table("out_100k_10k_1kb.pos-1", header=T)
#make allele 1 is the minor allele, 2 is the common allele
temp.idx <- snp$FREQ1 > snp$FREQ2
temp.freq <- snp$FREQ2
snp$FREQ2[temp.idx] <- snp$FREQ1[temp.idx]
snp$FREQ1[temp.idx] <- temp.freq[temp.idx]
#also change the genotype file
haplotype[,which(temp.idx==T)+2] <- 3 - haplotype[,which(temp.idx==T)+2]

#allele frequency
nrow(snp) #total number of snp
sum(snp$FREQ1 < 0.05) # number of snp with f < 0.05
sum(snp$FREQ1 < 0.01) # number of snp with f < 0.01
sum(snp$FREQ1 == 0.0001) # number of singletons

##assign risk variants and the corresponding effect size (proportional to allele frequency)
# null <- FALSE
n_haplo <- 10000
n_snp <- ncol(haplotype)-2
prevalence <- p_dis
b0_sqrt <- sqrt(prevalence)   #baseline

#set up causual SNPs
#generate risk haplotypes
# risk.variant.id <- c(3, 8,19,21,23,27,44,47,49,50)
risk.variant.id <- c(risk.variant) #2 for common 7 for rare 39 for super rare
risk.haplo.id <- which(apply(2-as.matrix(haplotype[, risk.variant.id+2]), 1, sum)>0)
(risk.haplo.f <- mean(apply(2-as.matrix(haplotype[, risk.variant.id+2]), 1, sum)>0)) #carrier haplotype frequency
haplotype.risk <- rep(1, length=nrow(haplotype))
#assign mean relative risk calculate the haplotype variants p(A|h)
haplotype.risk[risk.haplo.id] <- r
mean(haplotype.risk[risk.haplo.id]) #mean relative risk
haplotype.risk <<- haplotype.risk*b0_sqrt

##gene drop simulation for two generations
family_strct.2g2c <- data.frame(family=c(1,1,1,1), person=c(1,2,3,4), father=c(0,0,1,1), 
                                mother=c(0,0,2,2), sex=c(1,2,1,1), affect=c(1,2,2,2)) #1=male, 2=female, 1=unaffected, 2=affected


rep.idx <<- 1
sim_result <- replicate(n_rep, {
  print(rep.idx)
  rep.idx <<- rep.idx + 1
  

  family_generated_2g2c <<- gene_family(family_strct=family_strct.2g2c, n=n_family, haplotype.risk=haplotype.risk) 
  
  #remove the founder from the data
  family_generated_2c <- family_generated_2g2c
  family_generated_founder <- list()
  temp.idx <- which(family_generated_2g2c$data_family$person %in% 1:2) #take out founders
  family_generated_founder$data_family <- family_generated_2g2c$data_family[temp.idx, ]
  family_generated_founder$tran_vec <- family_generated_2g2c$tran_vec[temp.idx, ]
  family_generated_2c$data_family <- family_generated_2g2c$data_family[-temp.idx, ]
  family_generated_2c$tran_vec <- family_generated_2g2c$tran_vec[-temp.idx, ]

  #check if founder's haplotype carries any variant's with f < 0.1
  if(length(risk.variant.id)==1 & risk.variant.id[1] < 1) { #support allele frequency or list of snps
    snp2look.idx <-  which(snp$FREQ1 < risk.variant.id) # snp to look for
  } else(snp2look.idx <-  risk.variant.id)
  
  #disguitish those siblings with four founder's haplotype
  #calculate the allele frequency
  data <- family_generated_2c
  n_family <- max(data$data_family$family)
  n_family_member <- table(data$data_family$family)
  carrier.list <- list() #to store if the haplotype carrying a risk variant on multiple affected
  haplo.unique.count <- list()
  carrier.count.2 <- 0 
  chromosome.count.2 <- 0
  carrier.count.3 <- 0 
  chromosome.count.3 <- 0
  carrier.count.4 <- 0 
  chromosome.count.4 <- 0
  carrier.count.all <- 0 
  chromosome.count.all <- 0
  for(family.idx in 1:n_family) {
    current_row=sum(n_family_member[1:family.idx]) - n_family_member[family.idx]
    #assign a suitable transmission, not trivial to code up the algorithm assume is known for now
    tran_vec <- data$tran_vec[(current_row+1):(current_row+n_family_member[family.idx]),]
    ##tally the (un)ambiguous carrier haplotype
    h1 <- data$data_family[(current_row+1):(current_row+n_family_member[family.idx]),7:(6+n_snp)] #the first haplotype
    h2 <- data$data_family[(current_row+1):(current_row+n_family_member[family.idx]),-c(1:(6+n_snp))] #the second haplotype
    #observed allele count for each individual
    carrier <- c(0,0,0,0)
    for(a in 1:n_family_member[family.idx]) {
      idx.h1 <- match(tran_vec[a,"h1"], c("A", "B", "C", "D"))
      if(h1[a, snp2look.idx]==1) carrier[idx.h1] <- carrier[idx.h1] + 1
      idx.h2 <- match(tran_vec[a,"h2"], c("A", "B", "C", "D"))
      if(h2[a, snp2look.idx]==1) carrier[idx.h2] <- carrier[idx.h2] + 1
    }
    #the number of each unique haplotype occured on affecteds
    haplo.unique <- length(unique(as.vector(as.matrix(tran_vec[ ,c("h1","h2")]))))
    
    carrier.list[[family.idx]] <- carrier
    haplo.unique.count[[family.idx]] <- haplo.unique
    if(haplo.unique==2) {
      carrier.count.2 <- carrier.count.2 + sum(carrier>=1)
      chromosome.count.2 <- chromosome.count.2 + haplo.unique
    }
    if(haplo.unique==3) {
      carrier.count.3 <- carrier.count.3 + sum(carrier>=1)
      chromosome.count.3 <- chromosome.count.3 + haplo.unique
    }
    if(haplo.unique==4) {
      carrier.count.4 <- carrier.count.4 + sum(carrier>=1)
      chromosome.count.4 <- chromosome.count.4 + haplo.unique
    }
    if(haplo.unique>0) {
      carrier.count.all <- carrier.count.all + sum(carrier>=1)
      chromosome.count.all <- chromosome.count.all + haplo.unique
    }
  }
  #the allele frequency
  (founder.freq.offspring.2 <- carrier.count.2/chromosome.count.2)
  (founder.freq.offspring.3 <- carrier.count.3/chromosome.count.3)
  (founder.freq.offspring.4 <- carrier.count.4/chromosome.count.4)
  (founder.freq.offspring.all <- carrier.count.all/chromosome.count.all)
  
  ##founder's frequency
  data <- family_generated_founder
  carrier.count <- 0 
  chromosome.count <- 0
  for(family.idx in 1:n_family) {
    current_row=2*(family.idx-1)
    #assign a suitable transmission, not trivial to code up the algorithm assume is known for now
    tran_vec <- data$tran_vec[(current_row+1):(current_row+2),]
    ##tally the (un)ambiguous carrier haplotype
    h1 <- data$data_family[(current_row+1):(current_row+2),7:(6+n_snp)] #the first haplotype
    h2 <- data$data_family[(current_row+1):(current_row+2),-c(1:(6+n_snp))] #the second haplotype
    #observed allele count for each individual
    carrier <- 0
    for(a in 1:2) {
      if(h1[a, snp2look.idx]==1) carrier <- carrier + 1
      if(h2[a, snp2look.idx]==1) carrier <- carrier + 1
    }
    carrier.count <- carrier.count + carrier
    chromosome.count <- chromosome.count + 4
  }
  #the allele frequency
  (founder.freq.founder <- carrier.count/chromosome.count)
  
  c(founder.freq.offspring.2, founder.freq.offspring.3, founder.freq.offspring.4, 
    founder.freq.offspring.all, founder.freq.founder)
  
#   #test based on new imputation framework
#   result.trap.impute.2c <- family.test.nofounder.impute(data=family_generated_2c, f=risk.variant.id, method="trap")
#   result.trap.impute.2c.null <- family.test.nofounder.impute(data=family_generated_2c, f=risk.variant.id, method="trap", null.maf=snp$FREQ1[risk.variant.id])
#   result.trap.impute.noaggregate.2c <- family.test.nofounder.impute.noaggregate(data=family_generated_2c, f=risk.variant.id, method="trap")
#   result.trap.impute.noaggregate.2c.null <- family.test.nofounder.impute.noaggregate(data=family_generated_2c, f=risk.variant.id, method="trap", null.maf=snp$FREQ1[risk.variant.id])
#   result.trafic.impute.2c <- family.test.nofounder.impute(data=family_generated_2c, f=risk.variant.id, method="trafic")
#   result.trafic.impute.2c.null <- family.test.nofounder.impute(data=family_generated_2c, f=risk.variant.id, method="trafic", null.maf=snp$FREQ1[risk.variant.id])
#   result.trafic.impute.noaggregate.2c <- family.test.nofounder.impute.noaggregate(data=family_generated_2c, f=risk.variant.id, method="trafic")
#   result.trafic.impute.noaggregate.2c.null <- family.test.nofounder.impute.noaggregate(data=family_generated_2c, f=risk.variant.id, method="trafic", null.maf=snp$FREQ1[risk.variant.id])
#   
#   
#   ##below is the baseline power to compare with to see how much power lost
#   result.trap.nofounder.2c <- family.test.nofounder(data=family_generated_2c, f=risk.variant.id)
#   result.trafic.ext.2c <- family.test.trafic.ext(data=family_generated_2c, f=risk.variant.id)
#    
#   
#   #only report p.value
#   c(result.trap.impute.2c=result.trap.impute.2c, 
#     result.trap.impute.2c.null=result.trap.impute.2c.null,
#     result.trap.impute.noaggregate.2c=result.trap.impute.noaggregate.2c,
#     result.trap.impute.noaggregate.2c.null=result.trap.impute.noaggregate.2c.null,
#     result.trafic.impute.2c=result.trafic.impute.2c,
#     result.trafic.impute.2c.null=result.trafic.impute.2c.null,
#     result.trafic.impute.noaggregate.2c=result.trafic.impute.noaggregate.2c,
#     result.trafic.impute.noaggregate.2c.null=result.trafic.impute.noaggregate.2c.null,
#     result.trap.nofounder.2c=result.trap.nofounder.2c[7],
#     result.trafic.ext.2c=result.trafic.ext.2c)
})
## Write out your results to a csv file
result.df <- as.data.frame(t(sim_result))
colnames(result.df) <- c("founder.freq.offspring.2", "founder.freq.offspring.3", "founder.freq.offspring.4", 
                         "founder.freq.offspring.all", "founder.freq.founder")
result.df <- cbind(seed,r,p_dis,risk.variant,risk.haplo.f,n_family,result.df)
write.csv(result.df, paste("res_",r,"_",n_family,"_",seed,".csv",sep=""), row.names=FALSE)
## R.miSniam
## End(Not run)
## Not run:

##commands to run jobs in parallel
##passed in from the command prompt.
parseCommandArgs()
## Will disply the default values on the first run,
##initialize
seed=1000
#number of replications
n_rep=100
#number of family
n_family=1000
#prevalence
p_dis=0.3
# print(null) #null=FALSE
#family structure
# print(family_strct) #family_strct='family_strct.2g2c'

print(getwd())

##command line
parallel <- function(...) {
  names.args <- names(list(...))
  value.args <- as.vector(list(...))
  ##commands to run
  for(i in 1:1) {
    rfile="mainSim.R"
    cmd <- paste("R --vanilla --args seed ", seed, " ", paste(names.args, value.args, collapse=" "),
                 " < ", rfile, " > ", "mainSim_", paste(value.args, collapse="_"), ".Rout", seed, " 2>&1", sep="")
    print(cmd)
    writeLines(cmd, fileConn) #write jobs to text
    
    #add seed number
    seed<<-seed+1
  }
}
##clean-up & initialization
system("rm *.csv")
system("rm *.Rout*")
system("rm *.out*")
fileConn<-file("jobs.txt", "w")

##create your jobs here
for(i in seq(1,2, by=0.1)) {
  parallel(n_rep=n_rep, r=i, n_family=n_family, p_dis=p_dis, risk.variant=2) #common
}
for(i in seq(1,2, by=0.1)) {
  parallel(n_rep=n_rep, r=i, n_family=n_family, p_dis=p_dis, risk.variant=7) #rare
}
for(i in seq(1,2, by=0.1)) {
  parallel(n_rep=n_rep, r=i, n_family=n_family, p_dis=p_dis, risk.variant=39) #super rare
}


###################################################
#Load results with 3g and check the estiamted allele frequency
#three versions -- command and rare and super rare #2 for common, 7 for rare, 39 for super rare
result <- read.csv("2015-06-03 imputation framework for TRAP to fix common variant problem test on sibpair.csv", header=T)
result <- result %>% melt(id.vars=c("seed", "r", "p_dis","risk.variant","risk.haplo.f","n_family"), variable.name="method", value.name="freq.est")
result.plot <- result %>% group_by(risk.variant, r, method) %>% 
  summarise(n=n(), mu=mean(freq.est), se=sd(freq.est))

#TRAP
pd <- position_dodge(0.05)
filter(result.plot, risk.variant==39 & method %in% c("founder.freq.offspring.4", "founder.freq.founder")) %>% ggplot(aes(x=r, y=mu, group=method, col=method)) +
  #   geom_point(size=3, alpha=1) +
  geom_line(size=1.2, alpha=0.7, position=pd) +
  geom_point(size=1.2, position=pd) +
  geom_errorbar(aes(ymin=mu-se, ymax=mu+se), position=pd) +
#   ggtitle("f=0.202, consider effect size of risk haplotypes, TRAP") +
#   ggtitle("f=0.0178, consider effect size of risk haplotypes, TRAP") +
  ggtitle("f=0.0039, consider effect size of risk haplotypes, TRAP") +
  labs(x="relative risk r") +
  scale_y_continuous(limits=c(0,0.03)) +
  theme_gray(base_size = 20)
