source("function.r")
#need to take in parameter
## passed in from the command prompt.
parseCommandArgs()
## Will disply the default values on the first run,
print(seed)
print(n_rep)
print(r)
print(n_family) #number of family
print(p_dis) #prevelance

## ... your simualtion codek
##the founder's haplotypes
#import the haplotypes generated by cosi
haplotype <- read.table("out_100k_10k_1kb.hap-1", header=F)
colnames(haplotype) <- c("HAP", "CHROM", paste("SNP", 1:(ncol(haplotype)-2), sep=""))
snp <-read.table("out_100k_10k_1kb.pos-1", header=T)
#make allele 1 is the minor allele, 2 is the common allele
temp.idx <- snp$FREQ1 > snp$FREQ2
temp.freq <- snp$FREQ2
snp$FREQ2[temp.idx] <- snp$FREQ1[temp.idx]
snp$FREQ1[temp.idx] <- temp.freq[temp.idx]
#also change the genotype file
haplotype[,which(temp.idx==T)+2] <- 3 - haplotype[,which(temp.idx==T)+2]

#allele frequency
nrow(snp) #total number of snp
sum(snp$FREQ1 < 0.05) # number of snp with f < 0.05
sum(snp$FREQ1 < 0.01) # number of snp with f < 0.01
sum(snp$FREQ1 == 0.0001) # number of singletons

##assign risk variants and the corresponding effect size (proportional to allele frequency)
# null <- FALSE
n_haplo <- 10000
n_snp <- ncol(haplotype)-2
prevalence <- p_dis
b0_sqrt <- sqrt(prevalence)   #baseline

#set up causual SNPs
#generate risk haplotypes
risk.variant.id <- c(3, 8,19,21,23,27,44,47,49,50)
risk.haplo.id <- which(apply(2-haplotype[, risk.variant.id+2], 1, sum)>0)
(rish.haplo.f <- mean(apply(2-haplotype[, risk.variant.id+2], 1, sum)>0)) #carrier haplotype frequency
haplotype.risk <- rep(1, length=nrow(haplotype))
#assign mean relative risk calculate the haplotype variants p(A|h)
haplotype.risk[risk.haplo.id] <- r
mean(haplotype.risk[risk.haplo.id]) #mean relative risk
haplotype.risk <<- haplotype.risk*b0_sqrt

##gene drop simulation for two generations
family_strct.2g2c <- data.frame(family=c(1,1,1,1), person=c(1,2,3,4), father=c(0,0,1,1), 
                                mother=c(0,0,2,2), sex=c(1,2,1,1), affect=c(1,2,2,2)) #1=male, 2=female, 1=unaffected, 2=affected

rep.idx <<- 1
sim_result <- replicate(n_rep, {
  print(rep.idx)
  rep.idx <<- rep.idx + 1
  
  family_generated_2g2c <- gene_family(family_strct=family_strct.2g2c, n=n_family, haplotype.risk=haplotype.risk) 
  
  result.trap.2g2c <- family.test(data=family_generated_2g2c, f=risk.variant.id)
  result.trafic.ext.2g2c <- family.test.trafic.ext(data=family_generated_2g2c, f=risk.variant.id)
  result.trafic.pairwise.2g2c <- trafic.pairwise.test(data=family_generated_2g2c, f=risk.variant.id)$p.value
  
  #only report p.value
  c(result.trap.2g2c[7], result.trafic.ext.2g2c, result.trafic.pairwise.2g2c)
})
## Write out your results to a csv file
result.df <- as.data.frame(t(sim_result))
colnames(result.df) <- c("result.trap.2g2c","result.trafic.ext.2g2c","result.trafic.pairwise.2g2c")
result.df <- cbind(seed,r,result.df)
write.csv(result.df, paste("res_",r,"_",n_family,"_",seed,".csv",sep=""), row.names=FALSE)

#plot for the power results adding omnibus test
result <- read.csv("2015-04-20 TRAP, TRAFIC, TRAFIC_pairwise power comparison.csv", header=T)
result <- melt(result, id.vars=c("seed", "r"), variable.name="method", value.name="p.value")
result.power <- result %>% group_by(r, method) %>% 
  summarise(n=n(), power=mean(p.value<0.05))


ggplot(result.power, aes(x=r, y=power, col=method)) +
  #   geom_point(size=3, alpha=1) +
  geom_line(size=1.2, alpha=0.7) +
  ggtitle("f=0.0102, consider effect size of risk haplotypes") +
  labs(x="relative risk r") +
  theme_gray(base_size = 20) +
  scale_color_discrete(name="Method", 
                       breaks=c("result.trap.2g2c", "result.trafic.ext.2g2c", "result.trafic.pairwise.2g2c"),
                       labels=c("TRAP", "TRAFIC_ext", "TRAFIC_pairwise"))


